File: solution/kotlin/day19/src/main/kotlin/blog/Blog.kt
package blog

import arrow.core.Either
import arrow.core.Either.Left
import arrow.core.Either.Right
import java.time.LocalDate

class Article(
    private val name: String,
    private val content: String,
    val comments: List<Comment> = mutableListOf()
) {
    private fun addComment(
        text: String,
        author: String,
        creationDate: LocalDate
    ): Either<Error, Article> = Comment(text, author, creationDate)
        .let { comment ->
            return when (comments.contains(comment)) {
                true -> Left(Error("This comment already exists in this article"))
                else -> Right(Article(name, content, comments + comment))
            }
        }

    fun addComment(text: String, author: String): Either<Error, Article> = addComment(text, author, LocalDate.now())
}

data class Comment(val text: String, val author: String, val creationDate: LocalDate)
data class Error(val message: String)

File: exercise/ts/day19/src/blog.ts
export class Article {
    private constructor(private readonly name: string,
                        private readonly content: string,
                        private readonly comments: Comment[]) {
    }

    public static create = (name: string, content: string): Article => new Article(name, content, []);

    public addComment = (text: string, author: string): Article =>
        this.addCommentWithCreationDate(text, author, new Date());

    public getComments = (): Comment[] => this.comments;

    private addCommentWithCreationDate = (text: string, author: string, creationDate: Date): Article => {
        const comment = new Comment(text, author, creationDate);

        if (this.comments.some(c => c.isSame(comment))) {
            throw new Error("Comment already exists");
        }
        return new Article(this.name, this.content, [...this.comments, comment]);
    };
}

export class Comment {
    constructor(
        public readonly text: string,
        public readonly author: string,
        public readonly creationDate: Date
    ) {
    }

    isSame = (otherComment: Comment) => this.text === otherComment.text
        && this.author === otherComment.author
        && this.creationDate.getDate() === otherComment.creationDate.getDate();
}
